conngen/conngen.cpp

```diff
@@ -37,185 +34,312 @@
 // Includes from sli:
 #include "token.h"
 
+namespace nest
+{
+
 void
-nest::cg_connect( nest::ConnectionGeneratorDatum& cg,
-  const index source_id,
-  const index target_id,
+cg_connect( ConnectionGeneratorDatum& cg,
+  const GIDCollection& source_gids,
+  const GIDCollection& target_gids,
   const DictionaryDatum& params_map,
   const Name& synmodel_name )
 {
-  Subnet* sources =
-    dynamic_cast< Subnet* >( kernel().node_manager.get_node( source_id ) );
-  if ( sources == NULL )
-  {
-    LOG( M_ERROR, "CGConnect_cg_i_i_D_l", "sources must be a subnet." );
-    throw SubnetExpected();
-  }
-  if ( not sources->is_homogeneous() )
-  {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "sources must be a homogeneous subnet." );
-    throw BadProperty();
-  }
-  if ( dynamic_cast< Subnet* >( *sources->local_begin() ) )
+  const Token synmodel =
+    kernel().model_manager.get_synapsedict()->lookup( synmodel_name );
+  if ( synmodel.empty() )
   {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "Only 1-dim subnets are supported as sources." );
-    throw BadProperty();
+    throw UnknownSynapseType( synmodel_name.toString() );
   }
+  const index synmodel_id = static_cast< index >( synmodel );
+
+  RangeSet source_ranges, target_ranges;
+
+  cg_get_ranges( source_ranges, source_gids );
+  cg_get_ranges( target_ranges, target_gids );
+  cg_set_masks( cg, source_ranges, target_ranges );
+
+  cg->start();
 
-  Subnet* targets =
-    dynamic_cast< Subnet* >( kernel().node_manager.get_node( target_id ) );
-  if ( targets == NULL )
+  int source, target, num_parameters = cg->arity();
+  if ( num_parameters == 0 )
   {
-    LOG( M_ERROR, "CGConnect_cg_i_i_D_l", "targets must be a subnet." );
-    throw SubnetExpected();
+    // connect source to target
+    while ( cg->next( source, target, NULL ) )
+    {
+      // No need to check for locality of the target node, as the mask
+      // created by cg_set_masks() only contain local nodes.
+      Node* const target_node =
+        kernel().node_manager.get_node( target_gids[ target ] );
+      const thread target_thread = target_node->get_thread();
+      kernel().connection_manager.connect(
+        source_gids[ source ], target_node, target_thread, synmodel_id );
+    }
   }
-  if ( not targets->is_homogeneous() )
+  else if ( num_parameters == 2 )
   {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "targets must be a homogeneous subnet." );
-    throw BadProperty();
+    if ( not params_map->known( names::weight )
+      or not params_map->known( names::delay ) )
+    {
+      throw BadProperty(
+        "The parameter map has to contain the indices of weight and delay." );
+    }
+
+    long w_idx = ( *params_map )[ names::weight ];
+    long d_idx = ( *params_map )[ names::delay ];
+    std::vector< double > params( 2 );
+
+    // connect source to target with weight and delay
+    while ( cg->next( source, target, &params[ 0 ] ) )
```

Comment:
Is it really safe to use `&params[0]` here and assume contiguous memory layout for data in a `vector`?  I think the standard does not absolutely demand it; a C-style array would be safe.